File: csharp/0150-evaluate-reverse-polish-notation.cs
public partial class Solution {
    public int EvalRPN(string[] tokens) {

        var stack = new List<int>();

        foreach (var token in tokens) {
            if (token is not ("+" or "-" or "*" or "/")) {
                stack.Add(Convert.ToInt32(token));
                continue;
            }

            var num2 = stack.Last();
            stack.RemoveAt(stack.Count - 1);
            var num1 = stack.Last();
            stack.RemoveAt(stack.Count - 1);

            switch (token) {
                case "+":
                    stack.Add(num1 + num2);
                    break;
                case "-":
                    stack.Add(num1 - num2);
                    break;
                case "*":
                    stack.Add(num1 * num2);
                    break;
                case "/":
                    stack.Add(num1 / num2);
                    break;
            }
        }

        return stack[0];
    }
}

File: go/0150_evaluate_reverse_polish_notation.go
package main

// time: O(n)
// space: O(n)

import "strconv"

func evalRPN(tokens []string) int {
	stack := make([]int, 0)

	for _, token := range tokens {

		if token != "+" && token != "-" && token != "*" && token != "/" {
			num, _ := strconv.Atoi(token)
			stack = append(stack, num)
			continue
		}

		num1, num2 := stack[len(stack)-2], stack[len(stack)-1]
		stack = stack[:len(stack)-2]

		switch token {
		case "+":
			stack = append(stack, num1+num2)
			break
		case "-":
			stack = append(stack, num1-num2)
			break
		case "*":
			stack = append(stack, num1*num2)
			break
		case "/":
			stack = append(stack, num1/num2)
			break
		}
	}

	return stack[0]
}
